src/User/UserPassword.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\User;
+
+use Yiisoft\Security\PasswordHasher;
+use Yiisoft\Validator\Rule\HasLength;
+use Yiisoft\Validator\Rule\Required;
+use Yiisoft\Validator\SimpleRuleHandlerContainer;
+use Yiisoft\Validator\Validator;
+
+final class UserPassword
+{
+    private string $passwordHash;
+    private PasswordHasher $hasher;
+
+    private function __construct(private string $value)
+    {
+        $this->hasher = new PasswordHasher();
+        $this->passwordHash = $this->hasher->hash($this->value);
+    }
+
+    public function getHash(): string
+    {
+        return $this->passwordHash;
+    }
+
+    public function isEqualHash(string $hash): bool
+    {
+        return $this->hasher->validate($this->value, $hash);
+    }
+
+    /**
+     * @throw UserPasswordException
+     */
+    public static function createNew(string $password): UserPassword
+    {
+        $validator = new Validator(new SimpleRuleHandlerContainer());
```

Comment:
Assetions isn't what validator was meant for. I think that should stay at input validation level.